LCJava/src/main/java/tw/TrainRoute.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    public RouteResult searchPathMaxStops(String start, String end, int maxStops) {
+        int minStops = 1;
+        return this.searchPathByStops(start, end, minStops, maxStops);
+    }
+
+    public RouteResult searchPathExactStops(String start, String end, int exactStops) {
+        return this.searchPathByStops(start, end, exactStops, exactStops);
+    }
+
+    private RouteResult searchPathByStops(String start, String end, int minStops, int maxStops) {
+        List<String> result = new ArrayList<>();
+        List<String> path = new ArrayList<>();
+        path.add(start);
+        searchPathByStopsHelper(start, end, path, 0, minStops, maxStops, result);
+        if (result.size() > 0 ) {
+            return new RouteResult(result.size(), ResultMessage.SUCCESS);
+        }
+        return new RouteResult(-1, ResultMessage.FAILURE);
+    }
+
+    private void searchPathByStopsHelper(String start, String end, List<String> path, int currStops, int minStops, int maxStops, List<String> result) {
+        if (currStops > maxStops) {
+            return;
+        }
+        if (start.equals(end) && currStops >= minStops)  {
+            result.add(String.join("", path));
+        }
+        for (Map.Entry<String, Integer> entry: this.graph.get(start).entrySet()) {
+            String next = entry.getKey();
+            path.add(next);
+            searchPathByStopsHelper(next, end, path,currStops + 1, minStops, maxStops, result);
+            path.remove(path.size() - 1);
+        }
+    }
+
+    public RouteResult searchPathMaxDistance(String start, String end, int maxDistance) {
+        return this.searchPathByDistance(start, end, 1, maxDistance);
+    }
+
+    private RouteResult searchPathByDistance(String start, String end, int minDistance, int maxDistance ) {
+        List<String> result = new ArrayList<>();
+        List<String> path = new ArrayList<>();
+        path.add(start);
+        searchPathByDistanceHelper(start, end, path, 0, minDistance, maxDistance, result);
+        if (result.size() > 0 ) {
+            return new RouteResult(result.size(), ResultMessage.SUCCESS);
+        }
+        return new RouteResult(-1, ResultMessage.FAILURE);
+    }
+
+    private void searchPathByDistanceHelper(String start, String end, List<String> path, int distanceSum, int minDistance, int maxDistance, List<String> result ) {
+        if (distanceSum >= maxDistance) {
+            return;
+        }
+        if (start.equals(end) && distanceSum >= minDistance) {
+            result.add(String.join("", path));
+        }
+        for (Map.Entry<String, Integer> entry: this.graph.get(start).entrySet()) {
+            String next = entry.getKey();
+            path.add(next);
+            searchPathByDistanceHelper(next, end, path, distanceSum + entry.getValue(), minDistance, maxDistance, result);
+            path.remove(path.size() -1);
+        }
+    }
+
+    public RouteResult searchPathShortestDistance(String start, String end) {
```

Comment:
it's not a good idea to mix public/private methods. You can group methods of different access together for better readability.